lib/src/secure_config.rs

```diff
@@ -0,0 +1,571 @@
+        Self::new(repo_dir, "repo-config-id", "config.toml")
+    }
+
+    /// Creates a secure config for a workspace. Takes the .jj directory.
+    pub fn new_workspace(workspace_dir: PathBuf) -> Self {
+        Self::new(
+            workspace_dir,
+            "workspace-config-id",
+            "workspace-config.toml",
+        )
+    }
+
+    // If path_to_bytes failed, jj should not stop working.
+    // Instead, we simply treat all non-utf8 paths as empty.
+    fn encoded_path(&self) -> &[u8] {
+        path_to_bytes(&self.repo_dir).ok().unwrap_or_default()
+    }
+
+    fn generate_config_id(&self, rng: &RefCell<ChaCha20Rng>) -> String {
+        encode_hex(
+            &(0..CONFIG_ID_BYTES)
+                .map(|_| rng.borrow_mut().random::<u8>())
+                .collect::<Vec<_>>(),
+        )
+    }
+
+    fn generate_config(
+        &self,
+        config_dir: &Path,
+        config_id: &str,
+        content: &[u8],
+        metadata: &ConfigMetadata,
+    ) -> Result<PathBuf, SecureConfigError> {
+        let config_dir = config_dir.join(config_id);
+        let config_path = config_dir.join(CONFIG_FILE);
+        fs::create_dir_all(&config_dir).context(&config_dir)?;
+        self.update_metadata(&config_dir, metadata)?;
+        if !content.is_empty() {
+            fs::write(&config_path, content).context(&config_path)?;
+        }
+
+        // Write the config ID atomically. A half-formed config ID would be very bad.
+        atomic_write(
+            &self.repo_dir.join(self.config_id_name),
+            config_id.as_bytes(),
+        )?;
+        Ok(config_path)
+    }
+
+    fn generate_initial_config(
+        &self,
+        config_dir: &Path,
+        config_id: &str,
+    ) -> Result<(PathBuf, ConfigMetadata), SecureConfigError> {
+        let metadata = ConfigMetadata {
+            path: self.encoded_path().to_vec(),
+        };
+        let path = self.generate_config(config_dir, config_id, &[], &metadata)?;
+        Ok((path, metadata))
+    }
+
+    fn update_metadata(
+        &self,
+        config_dir: &Path,
+        metadata: &ConfigMetadata,
+    ) -> Result<(), SecureConfigError> {
+        let metadata_path = config_dir.join(METADATA_FILE);
+        atomic_write(&metadata_path, &metadata.encode_to_vec())?;
+        Ok(())
+    }
+
+    /// Validates that the metadata path matches the repo path.
+    /// If there's a mismatch, takes appropriate action.
+    /// Returns the updated config dir and metadata.
+    fn handle_metadata_path(
+        &self,
+        rng: &RefCell<ChaCha20Rng>,
+        root_config_dir: &Path,
+        config_dir: PathBuf,
+        mut metadata: ConfigMetadata,
+    ) -> Result<(PathBuf, ConfigMetadata), SecureConfigError> {
+        let (encoded, want) = if let Ok(encoded) = path_to_bytes(&self.repo_dir) {
+            (encoded, Some(self.repo_dir.as_path()))
+        } else {
+            Default::default()
+        };
+        let got = if metadata.path.is_empty() {
```

Comment:
Perhaps, we can mark the proto field as `optional bytes` and use `Option<Vec<u8>>`? That will help prevent misuse of paths that couldn't be encoded to bytes.